internal/pkg/deploy/upload/asset/asset.go

```diff
@@ -71,98 +87,73 @@ func walkFn(source, dest string, recursive bool, reader afero.Fs, upload UploadF
 		if !ok {
 			return nil
 		}
-		file, err := reader.Open(path)
+
+		hash := sha256.New()
+		buf := &bytes.Buffer{}
+		file, err := u.FS.Open(path)
 		if err != nil {
-			return fmt.Errorf("open file on path %q: %w", path, err)
+			return fmt.Errorf("open %q: %w", path, err)
 		}
 		defer file.Close()
-		fileRel, err := filepath.Rel(source, path)
+
+		_, err = io.Copy(io.MultiWriter(buf, hash), file)
 		if err != nil {
-			return fmt.Errorf("get relative path for %q against %q: %w", path, source, err)
+			return fmt.Errorf("copy %q: %w", path, err)
 		}
-		key := filepath.Join(dest, fileRel)
-		url, err := upload(key, file)
+
+		// rel is "." when sourcePath == path
+		rel, err := filepath.Rel(sourcePath, path)
 		if err != nil {
-			return fmt.Errorf("upload file %q to destination %q: %w", path, key, err)
+			return fmt.Errorf("get relative path for %q against %q: %w", path, sourcePath, err)
 		}
-		*urlsPtr = append(*urlsPtr, url)
-		return nil
-	}
-}
 
-type filepathMatcher interface {
-	match(path string) (bool, error)
-}
-
-type reincludeMatcher string
+		dest := filepath.Join(destPath, rel)
+		if dest == "." { // happens when sourcePath is a file and destPath is unset
+			dest = info.Name()
```

Comment:
would this logic make sense for all kinds of `assets` (for example if we support lambdas in the future, then the `assets` might be the JS codes), or just the static site assets?

Reply:
lambdas are mostly different because they require a single zip file - if we were to support a lambda pattern, i would think it would be more as a containerized lambda, not directly uploading assets to s3 🤔 (something to think about in the future!!!)